src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -129,19 +129,34 @@ T dot(const SGVector<T>& a, const SGVector<T>& b)
 }
 
 /**
- * Method that computes the mean of vectors or matrices.
+ * Method that computes the mean of vectors or matrices composed of complex numbers.
  *
  * @param a SGVector or SGMatrix
  * @return the vector mean \f$\bar{a}_i\f$ or matrix mean \f$\bar{m}_{i,j}\f$
  */
-template <typename T, template <typename> class Container>
-typename linalg::int2float<T>::floatType mean(const Container<T>& a)
+template<typename T, template <typename> class Container>
+typename std::enable_if<std::is_same<T, complex128_t>::value, complex128_t>::type
+mean(const Container<T>& a)
 {
 	REQUIRE(a.size() > 0, "Vector/Matrix cannot be empty!\n");
 	return infer_backend(a)->mean(a);
 }
 
 /**
+ * Method that computes the mean of vectors or matrices composed of real numbers.
+ *
+ * @param a SGVector or SGMatrix
+ * @return the vector mean \f$\bar{a}_i\f$ or matrix mean \f$\bar{m}_{i,j}\f$
+ */
+template <typename T, template <typename> class Container>
+typename std::enable_if<!std::is_same<T, complex128_t>::value, float64_t>::type
+mean(const Container<T>& a)
+{
+	REQUIRE(a.size() > 0, "Vector/Matrix cannot be empty!\n");
+	return std::abs(infer_backend(a)->mean(a));
+}
+
```

Comment:
no you are getting this a bit wrong. no need for casting or anything, every interface (complex vs rest) should just call the right function. so in other words when you have:
```
template <typename T, template <typename> class Container>
complex128_t mean_impl(const Container<T>& a) const;
```

then it should actually call:
```
complex64_t sum_impl(Container<complex128_t>& a)
```
or use `enabled_if`, but don't enable this for *every* `T`.

you should not define a generic:
```
template <typename T, template <typename> class Container>
complex128_t mean_impl(const Container<T>& a) const;
```
it should only be defined for `T = complex128_t`.